pkg/controller/mysqlcluster/mysqlcluster_controller_test.go

```diff
@@ -17,65 +16,298 @@ limitations under the License.
+
+		recFn, requests = SetupTestReconcile(newReconciler(mgr))
+		Expect(add(mgr, recFn)).To(Succeed())
+
+		stop = StartTestManager(mgr)
+	})
+
+	AfterEach(func() {
+		close(stop)
+	})
+
+	Describe("when creating a new mysql cluster", func() {
+		var (
+			expectedRequest reconcile.Request
+			cluster         *api.MysqlCluster
+			secret          *core.Secret
+			components      clusterComponents
+		)
+
+		BeforeEach(func() {
+			name := fmt.Sprintf("cluster-%d", rand.Int31())
+			ns := "default"
+
+			expectedRequest = reconcile.Request{
+				NamespacedName: types.NamespacedName{Name: name, Namespace: ns},
+			}
+
+			secret = &core.Secret{
+				ObjectMeta: metav1.ObjectMeta{Name: "the-secret", Namespace: ns},
+				StringData: map[string]string{
+					"ROOT_PASSWORD": "this-is-secret",
+				},
+			}
+
+			cluster = &api.MysqlCluster{
+				ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: ns},
+				Spec: api.MysqlClusterSpec{
+					Replicas:   1,
+					SecretName: secret.Name,
+				},
+			}
+
+			components = []runtime.Object{
+				&apps.StatefulSet{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.StatefulSet),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.HeadlessSVC),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.MasterService),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.HealthyNodesService),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&core.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.ConfigMap),
+						Namespace: cluster.Namespace,
+					},
+				},
+				&policy.PodDisruptionBudget{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      cluster.GetNameForResource(api.PodDisruptionBudget),
+						Namespace: cluster.Namespace,
+					},
+				},
+			}
+
+			Expect(c.Create(context.TODO(), secret)).To(Succeed())
+
+		})
+
+		AfterEach(func() {
+			// manually delete all created resources because GC isn't enabled in the test controller plane
+			removeAllCreatedResource(c, components)
+			c.Delete(context.TODO(), secret)
+		})
+
+		It("should reconcile the statefulset", func() {
+			Expect(c.Create(context.TODO(), cluster)).To(Succeed())
+			defer c.Delete(context.TODO(), cluster)
+			Eventually(requests, timeout).Should(Receive(Equal(expectedRequest)))
+
+			sfsKey := types.NamespacedName{
+				Name:      cluster.GetNameForResource(api.StatefulSet),
+				Namespace: cluster.Namespace,
+			}
+			statefulSet := &apps.StatefulSet{}
+			Eventually(func() error {
+				return c.Get(context.TODO(), sfsKey, statefulSet)
+			}, timeout).Should(Succeed())
+
+			When("delete the statefulset", func() {
```

Comment:
Move this test in reconcile components. The test function for a component, should test that after we delete that component, the controller recreates it.